src/posix.c

```diff
@@ -182,15 +198,45 @@ int p_fallocate(int fd, off_t offset, off_t len)
 
 	/*
 	 * fcntl will often error when the file already exists; ignore
-	 * this error since ftruncate will also resize the file (although
+	 * this error since our fallback will also resize the file (although
 	 * likely slower).
 	 */
-	fcntl(fd, F_PREALLOCATE, &prealloc);
-
-	return ftruncate(fd, (offset + len));
+	if ((error = fcntl(fd, F_PREALLOCATE, &prealloc)) < 0)
+		goto fallback;
+#elif defined(GIT_USE_FALLOCATE)
+	do {
+		error = fallocate(fd, 0, offset, len);
+	} while (error < 0 && errno == EINTR);
```

Comment:
Can `fallocate` set `EAGAIN`?

Reply:
That's copy-pasted from https://github.com/libgit2/libgit2/pull/5087#issuecomment-495979590, actually. And isn't  `EAGAIN` usually async network stuff ? I'm scared now…